### PR TITLE
ANNOT: Detect wrong safe or unsafe impls E0199/E0200

### PR DESCRIPTION
**Pull request by farodin91**
#886

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 1]
```diff
@@ -311,6 +311,11 @@ class RsErrorAnnotator : Annotator, HighlightRangeExtension {
     private fun checkImpl(holder: AnnotationHolder, impl: RsImplItem) {
         val trait = impl.traitRef?.resolveToTrait ?: return
         val traitName = trait.name ?: return
+        if (impl.unsafe != null && trait.unsafe == null) {
+            holder.createErrorAnnotation(impl.traitRef!!, "Safe traits should not have unsafe implementations [E0199]")
```
Comment (alygin) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: We usually try to stick to the exact (or at least as close as we can) error messages that the compiler emits. In these cases they are "Implementing the trait \`Foo\` is not unsafe [E0199]" and "The trait \`Foo\` requires an \`unsafe impl\` declaration [E0200]". Could you, please, change them accordingly?

File: src/test/kotlin/org/rust/ide/annotator/RsErrorAnnotatorTest.kt [hunk 2]
```diff
@@ -860,4 +860,24 @@ class RsErrorAnnotatorTest : RsAnnotatorTestBase() {
             self::foo()
         }
     """)
+
+    fun `testE0199 Only safe impls for safe traits`() = checkErrors("""
```
Comment (alygin) [inline src/test/kotlin/org/rust/ide/annotator/RsErrorAnnotatorTest.kt]: Annotation tests in this file are ordered respectively to their error codes. Would you, please, move them to the right positions? It makes maintenance a bit easier.

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 3]
```diff
@@ -311,6 +311,11 @@ class RsErrorAnnotator : Annotator, HighlightRangeExtension {
...
+        } else if (impl.unsafe == null && trait.unsafe != null) {
+            holder.createErrorAnnotation(impl.traitRef!!, "Unsafe traits must have unsafe implementations [E0200]")
```
Comment (alygin) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: Let's declare `traitRef = impl.traitRef ?: return` at the very method beginning and get rid of these `!!`s.

**Comment by farodin91** [comment]
@alygin Resolved!

**Comment by alygin** [comment]
Thanks!

bors r+

**Pull request merged**